aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu

```diff
@@ -108,32 +112,36 @@ Tensor& addmm_out_sparse_csr_dense_cuda(
   }
 
   if (&result != &self) {
-    at::native::resize_output(result, self__sizes);
-    if (beta.toComplexDouble() != 0.0) {
-      at::native::copy_(result, *self_);
+    if (result.layout() == kStrided) {
+      at::native::resize_output(result, self__sizes);
+    } else {
+      at::native::resize_as_sparse_csr_(result, *self_);
     }
+    result.copy_(*self_);
   }
 
   IntArrayRef result_sizes = result.sizes();
   if ((result_sizes[0] == 0) || (result_sizes[1] == 0)) {
     return result;
   }
 
-  if (mat1._nnz() == 0) {
-    // By definition, when beta==0, values in self should be ignored. nans and infs
-    // should not propagate
+  if (mat1._nnz() == 0 && mat2.layout() == kStrided) {
+    // According to docs, when beta==0 values in self should be ignored. nans and infs should not propagate
     if (beta.toComplexDouble() == 0.) {
-      return result.zero_();
+      result.zero_();
+    } else {
+      result.mul_(beta);
     }
-    return at::mul_out(
-        result,
-        self,
-        at::native::scalar_tensor(
-            beta,
-            self.scalar_type(),
-            c10::nullopt /* layout */,
-            at::kCPU,
-            c10::nullopt /* pin_memory */));
+    return result;
+  }
+
+  if (mat2.is_sparse_csr() && (mat1._nnz() == 0 || mat2._nnz() == 0)) {
+    if (beta.toComplexDouble() == 0.) {
+      result.values().zero_();
```

Comment:
Why don't we also want to reset the column and row indices? Instead of setting the values to zero we could erase those.

Reply:
I didn't think about it. The current code is simpler, erasing the sparsity pattern might not be the desired behavior. Better to do it explicitly with some other functionality.

Reply:
I think this entirely depends on the subsequent operation. You could make the argument that filling values with zeros is wasteful (and it indeed causes another kernel launch). Not sure what we should do here. Maybe @ngimel has some advice?